Refactor for suggestion modal display

ankita240796 opened the pull request:
## Explanation

This PR refactors the suggestion modal display schema. The structure is as follows:

- A service for displaying suggestions in different views. Different types of suggestions will be displayed using same service. But for suggestion in different views there are separate services.
- A controller for displaying suggestions in different views. Different types of suggestions will be displayed using same contoller. But for suggestion in different views there are separate controllers.
- HTML templates for each pair of view and suggestion type.

Currently the PR has structure for only `edit_state_content` suggestion. Further changes will be made in fixing #5546

ankita240796 committed:
Added new service for suggestion modal display

ankita240796 committed:
Fix nits in develop

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into suggestion-view

ankita240796 committed:
Fix structure issues

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into suggestion-view

ankita240796 committed:
Fix minor nit

ankita240796 commented:
Hi @nithusha21, PTAL. Thanks!

nithusha21 commented on core/templates/dev/head/pages/base.html:
```diff
@@ -177,6 +177,8 @@ <h2>
     <script src="/templates/dev/head/i18n.js"></script>
 
     <script src="/templates/dev/head/pages/Base.js"></script>
+    <script src="/templates/dev/head/pages/suggestion_editor/ShowSuggestionModalForLearnerLocalViewService.js"></script>
```
Can we keep this sorted? This will go below the other js file then.

nithusha21 replied:
By the way, why are we importing this in base.html? Can we keep it in some html where it is more specific?

ankita240796 replied:
Done.

nithusha21 commented on core/templates/dev/head/pages/creator_dashboard/CreatorDashboard.js:
```diff
@@ -319,196 +321,15 @@ oppia.controller('CreatorDashboard', [
+        $scope.activeThread.suggestion.suggestionType,
+        {
+          activeThread: $scope.activeThread,
+          suggestionsToReviewList: $scope.suggestionsToReviewList,
```
Why do we need to pass this in?

ankita240796 replied:
The parameters passed here are required to handle a suggestion in the SuggestionService for creator view (https://github.com/oppia/oppia/pull/6093/files#diff-3cd0164785e4bfe2277b36148961a0c2R30). So, we need to pass them.

nithusha21 commented on core/templates/dev/head/pages/suggestion_editor/ShowSuggestionModalForCreatorView.js:
```diff
@@ -0,0 +1,122 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
```
It's 2019 :)

ankita240796 replied:
My bad, fixed this.

nithusha21 commented on core/templates/dev/head/pages/suggestion_editor/ShowSuggestionModalForCreatorViewService.js:
```diff
@@ -0,0 +1,133 @@
+            return canReviewActiveThread;
+          },
+          stateName: function() {
+            return activeThread.suggestion.stateName;
```
I think this value would be null always (could you check?). If I remember correctly, this field was removed. If so, we can delete this function (and other places where state_name is used).

ankita240796 replied:
The value is the state name for which the suggestion is provided.

nithusha21 replied:
While manually testing, did you run into this field being non-null at any point? (IIRC, this was removed from the backend, so I don't think this data would exist.)

nithusha21 replied:
Just this one question

ankita240796 replied:
Ah sorry @nithusha21, I missed this one. I actually just tested by adding `console.log(activeThread.suggestion.stateName)` after starting a local server and that produced a valid statename for which the suggestion was made.

Is there any other path which I should follow for testing this?

nithusha21 replied:
Oh that's interesting. I think it is fine as is. We can take a look at this independently.

ankita240796 replied:
Ok, Thanks!

nithusha21 reviewed (changes requested):
Hi @ankita240796, really sorry for the delay in reviewing. I really like where this PR is going. A couple of questions, and possibly cleanup (might have been existing problems). Also, a lot of code seems to be common between the 3 different controllers/services? Would a dedicated service to just make the backend requests be useful? Maybe we can encapsulate all the accept/reject/resubmit behaviour in a separate service... Would that be viable?

ankita240796 commented:
Hi @nithusha21, Sorry for the delay. I will update the PR by tomorrow, Thanks!

ankita240796 committed:
Resolve conflict

ankita240796 committed:
Change 2018 to 2019, include scripts in correct file

ankita240796 commented:
Hi @nithusha21, I have addressed the review comments, PTAL!

I am working on creating a dedicated service to remove common functions from service/controllers. I will make the changes by tonight. 

Thanks!

ankita240796 committed:
Add service to handle common code for suggestion modal display

ankita240796 commented:
Hi @nithusha21, I have added a dedicated service to handle common code for suggestion modal display. We may need to move more common code to the service but I could have a better understanding for doing that if you review the initial draft first. PTAL, Thanks!

ankita240796 commented:
Hi @nithusha21, PTAL. Thanks!

DubeySandeep commented:
@nithusha21, can you please review this PR!

@ankita240796, make sure you have followed [these UX guidelines](https://github.com/oppia/oppia/wiki/Oppia-UX-guidelines-&-rationales) for modal(s)
Thanks!.

nithusha21 commented:
I would need to test the user pathways for this (As this affects critical user journeys). Will do so, and complete a code review by tonight!

nithusha21 reviewed (approved):
Hi @ankita240796, sorry for the delay. This PR looks amazing! I just tested creation, and actioning on suggestions. Everything works great! 

Thanks a lot for the changes :) 

I have one question (follow up). Once answered I'll merge.
